lib/nrf9160_time/nrf9160_time.c

```diff
@@ -0,0 +1,284 @@
+	struct addrinfo *addr;
+	struct sntp_ctx sntp_ctx;
+
+	hints.ai_family = AF_INET;
+	hints.ai_socktype = SOCK_DGRAM;
+	hints.ai_protocol = 0;
+
+	err = net_getaddrinfo_addr_str(server, "123", &hints, &addr);
+	if (err) {
+		LOG_ERR("net_getaddrinfo_addr_str, error: %d", err);
+		return err;
+	}
+
+	err = sntp_init(&sntp_ctx, addr->ai_addr, addr->ai_addrlen);
+	if (err) {
+		goto freeaddr;
+	}
+
+	err = sntp_query(&sntp_ctx, timeout, time);
+	if (err) {
+		sntp_close(&sntp_ctx);
+		goto freeaddr;
+	}
+
+freeaddr:
+	freeaddrinfo(addr);
+	return err;
+}
+
+static int get_time_NTP_server(void)
+{
+	int err;
+	int i = 0;
+
+	while (i < ARRAY_SIZE(servers)) {
+		err =  sntp_request_time(servers[i].server,
+					 K_SECONDS(5), &sntp_time);
+		if (err) {
+			LOG_DBG("Not getting time from NTP server %s, error %d",
+				log_strdup(servers[i].server), err);
+			LOG_DBG("Trying another address...");
+			i++;
+			continue;
+		}
+
+		LOG_DBG("Got time response from NTP server %s",
+			log_strdup(servers[i].server));
+		time_aux.date_time_utc = (s64_t)sntp_time.seconds * 1000;
+		time_aux.last_date_time_update = k_uptime_get();
+		return 0;
+	}
+
+	LOG_ERR("Not getting time from any NTP server");
+
+	return -ENODATA;
+}
+
+static int check_current_time(void)
+{
+	if (time_aux.last_date_time_update == 0 ||
+	    time_aux.date_time_utc == 0) {
+		LOG_DBG("Date time never set");
+		return -ENODATA;
+	}
+
+	if ((k_uptime_get() - time_aux.last_date_time_update) >
+	    CONFIG_NRF9160_TIME_UPDATE_INTERVAL * 1000) {
+		LOG_DBG("Current date time too old");
+		return -ENODATA;
+	}
+
+	return 0;
+}
+
+static void update_new_date_time(void)
+{
+	int err;
+
+	while (true) {
+		k_sem_take(&fetch_time_sem, K_FOREVER);
+
+		LOG_DBG("Updating date time UTC...");
+
+		err = check_current_time();
+		if (err == 0) {
+			LOG_DBG("Time successfully obtained");
+			initial_valid_time = true;
+			continue;
+		}
+
+		LOG_DBG("Current time not valid");
+
+#if defined(CONFIG_NRF9160_TIME_MODEM)
+		LOG_DBG("Fallback on cellular network time");
+
+		err = get_time_cellular_network();
+		if (err == 0) {
+			LOG_DBG("Time from cellular network obtained");
+			initial_valid_time = true;
+			continue;
+		}
+
+		LOG_DBG("Not getting cellular network time");
+#endif
+
+		LOG_DBG("Fallback on NTP server");
+
+		err = get_time_NTP_server();
+		if (err == 0) {
+			LOG_DBG("Time from NTP server obtained");
+			initial_valid_time = true;
+			continue;
+		}
+
+		LOG_DBG("Not getting time from NTP server");
+	}
+}
+
+K_THREAD_DEFINE(time_thread, CONFIG_NRF9160_TIME_THREAD_SIZE,
+		update_new_date_time, NULL, NULL, NULL,
+		K_HIGHEST_APPLICATION_THREAD_PRIO, 0, K_NO_WAIT);
+
+static void nrf9160_time_handler(struct k_work *work)
+{
+	k_sem_give(&fetch_time_sem);
+
+	LOG_DBG("New date time update in: %d seconds",
+		CONFIG_NRF9160_TIME_UPDATE_INTERVAL);
+
+	k_delayed_work_submit(&time_work,
+			      K_SECONDS(CONFIG_NRF9160_TIME_UPDATE_INTERVAL));
+}
+
+void nrf9160_time_init(void)
```

Comment:
This sound reasonable. The problem is that the device would not be connected to internet (and no GPS fix), essentially meaning that it will always output a sequence of error message at boot. But if this is considered OK. Why not.

Reply:
The initialization is now done at zephyr boot. The library will not try to fetch time upon initialization. I also added a function ``date_time_update`` which can be used by the application to trigger an update cycle independent of the internal update routine. This function can be called as soon as the device is connected to network.